dcos-log/mesos/files/reader/read.go

```diff
@@ -103,58 +108,68 @@ func NewLineReader(client *http.Client, masterURL url.URL, agentID, frameworkID,
 			length = rm.offset
 		}
 
-		// TODO: split this function into smaller ones to improve the readability.
-		for {
-			ctx, cancel := context.WithTimeout(context.Background(), time.Second*3)
-			lines, delta, err := rm.read(ctx, offset, length, reverse)
-			if err != nil {
-				cancel()
-				return nil, err
-			}
+		err := calcOffset(offset, length, rm)
+		if err != nil && err != io.EOF {
+			return nil, err
+		}
+	}
 
+	// guard against negative offset
+	if rm.offset < 0 {
```

Comment:
should this be an error condition?

Reply:
no, because we use a chunk with a fixed size.
Suppose we have a file that has only 100 bytes. A chunk is 65536, so when we read backwards the may endup with weird offset set to `-65436`, but that is actually OK, that just means we need to read the file from the beginning (offset = 0)

Reply:
makes sense thanks